craft_application/services/linter.py

```diff
@@ -0,0 +1,268 @@
+# This file is part of craft-application.
+#
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""Orchestrates linter registration, ignore config and execution."""
+
+from __future__ import annotations
+
+import inspect
+from collections.abc import Iterable, Iterator
+from typing import TYPE_CHECKING, Any, cast
+
+from craft_cli import emit
+
+from craft_application import util
+from craft_application.lint.types import (
+    ExitCode,
+    IgnoreConfig,
+    IgnoreSpec,
+    LintContext,
+    LinterIssue,
+    Severity,
+    Stage,
+    should_ignore,
+)
+from craft_application.services import base
+
+if TYPE_CHECKING:
+    from pathlib import Path
+
+    from craft_application.application import AppMetadata
+    from craft_application.lint.base import AbstractLinter
+    from craft_application.services.service_factory import ServiceFactory
+
+
+class LinterService(base.AppService):
+    """Orchestrates linter registration, ignore config and execution."""
+
+    _class_registry: dict[Stage, list[type[AbstractLinter]]] = {
+        Stage.PRE: [],
+        Stage.POST: [],
+    }
+
+    def __init__(self, app: AppMetadata, services: ServiceFactory) -> None:
+        super().__init__(app, services)
+        self._ignore_cfg: IgnoreConfig = {}
+        self._issues: list[LinterIssue] = []
+
+    @classmethod
+    def register(cls, linter_cls: type[AbstractLinter]) -> None:
+        """Register a linter class for use by the service."""
+        if inspect.isabstract(linter_cls):
+            return
+        stage = getattr(linter_cls, "stage", None)
+        name = getattr(linter_cls, "name", None)
```

Comment:
This is something that `AbstractLinter` should be checking, not the linter service.